### PR TITLE
mds: optionally forbid to use standby for another fs as last resort

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/61599

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by trociny** [comment]
Alternatively, instead of a config option we could use a fs option, something like `ceph fs set <fs_name> allow_another_fs_standby false`. @batrick @vshankar what do you think?

If we still want a config option then I am also not sure if `mds_allow_another_fs_standby` is a good name. The option should be set for mon while the `mds_` prefix suggests it is for mds. Then may be `mon_mds_allow_another_fs_standby`?

**Comment by batrick** [comment]
> Alternatively, instead of a config option we could use a fs option, something like `ceph fs set <fs_name> allow_another_fs_standby false`. @batrick @vshankar what do you think?

This would be more appropriate as you may want different behavior for different FS.

**Comment by trociny** [comment]
> This would be more appropriate as you may want different behavior for different FS.

That is what I would like more too. Thanks.

Then the next question is about naming and the default behaviour. Ideally I would like we don't allow standby for another fs by default and if one wants it they should set the fs flag with the command:
```
ceph fs set <fs> allow_another_fs_standby true
```
(an alternative flag name could be `allow_standby_for_another_fs` or may be you have some other idea for the name?)

But this would mean the previous behaviour would change after upgrade (and I would like to backport this to all supported branches) and users may not be happy. So a safer approach would be to add `refuse_another_fs_standby` (`refuse_standby_for_another_fs`) flag instead, which users are supposed to set if they want the new behaviour:
```
ceph fs set <fs> refuse_another_fs_standby true
```
I am using `allow_` and `refuse_` prefixes because I see we already have flags with these prefixes, still if you have a better idea about the flag name I would be glad to know.

So, what do you think? I would be fine with any solution that allowed to enable this somehow.

**Comment by batrick** [comment]
I don't think we can change the default behavior. It's too surprising to users why standbys are not being promoted.

Name is good to me.

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
@trociny Needs update to doc in doc/cephfs/standby.rst

... and just to clarify, the default behaviour is still what it is right now (monitor can choose any standby MDS)?

**Comment by trociny** [comment]
updated to include the test and documentation change.

**Comment by trociny** [comment]
Running on teuthology the fs suite with `--filter failover`, just to check that the modified `join_fs_last_resort` test works:

https://pulpito.ceph.com/trociny-2023-06-25_06:00:43-fs-wip-mgolub-testing-distro-default-smithi/

 All tests passed.

**Comment by vshankar** [comment]
jenkins test api

File: doc/cephfs/standby.rst
Comment (dparmar18) [inline doc/cephfs/standby.rst:130]:
```suggestion
.. note::
   Configuring MDS file system affinity does not guarantee that
   ``standby-replay`` daemons are always selected before other standbys.
```

File: PendingReleaseNotes
Comment (batrick) [inline PendingReleaseNotes:230]:
```suggestion
* Introduced a new file system flag `refuse_standby_for_another_fs` that can be
set using the `fs set` command. This flag prevents using a standby for another
file system (join_fs = X) when standby for the current filesystem is not available.
Relevant tracker: https://tracker.ceph.com/issues/61599
```

File: qa/tasks/cephfs/test_failover.py
Comment (batrick) [inline qa/tasks/cephfs/test_failover.py:140]: Please make a separate, focused test.

File: src/mon/FSCommands.cc
Comment (batrick) [inline src/mon/FSCommands.cc:742]:
```suggestion
          ss << "set to refuse standby for another fs";
```

File: src/mon/FSCommands.cc
Comment (batrick) [inline src/mon/FSCommands.cc:744]: ditto

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise LGTM

**Comment by vshankar** [comment]
> otherwise LGTM

@trociny I'll include this in my integration run this week. Could you finish up the comments?

**Comment by trociny** [comment]
> @trociny I'll include this in my integration run this week. Could you finish up the comments?

@vshankar I am on vacation right now. Actually, I think I addressed @batrick's comments but I have not had a chance to try the updated test so I was reluctant to push it. Anyway I pushed it right now though it is still not tested.

**Comment by trociny** [comment]
The updated tests seem to work:

https://pulpito.ceph.com/trociny-2023-07-04_04:45:48-fs-wip-mgolub-testing-distro-default-smithi/

Note, wip-mgolub-tesing branch has different sha than wip-61599 just because I had to rebase it to force shaman build (the previous build failed for some reason).

**Comment by vshankar** [comment]
> > @trociny I'll include this in my integration run this week. Could you finish up the comments?
> 
> @vshankar I am on vacation right now. Actually, I think I addressed @batrick's comments but I have not had a chance to try the updated test so I was reluctant to push it. Anyway I pushed it right now though it is still not tested.

Sure, no problem. I'll have a look. Thx!

**Comment by vshankar** [comment]
jenkins retest this please

**Review by batrick — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-api/58129/

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
I'll pick this PR for integration testing in few days.

File: qa/tasks/cephfs/test_failover.py
Comment (dparmar18) [inline qa/tasks/cephfs/test_failover.py:183]: this is same as in `test_join_fs_last_resort`. This helper can be moved before the test cases start; thus removing redundancy.

File: src/mds/FSMap.cc
Comment (dparmar18) [reply src/mds/FSMap.cc:796]: usage of dot operator looks correct

**Review by rishabh-d-dave — APPROVED** [review]
Testing was successful - https://tracker.ceph.com/projects/cephfs/wiki/Main#24-Jul-2024

**Comment by rishabh-d-dave** [comment]
The PR is ready for merge but not until Ceph API tests also passes. I'll merge after as soon as it passes.

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-windows-pull-requests/25605/

**Comment by rishabh-d-dave** [comment]
jenkins test windows

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/43130/

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-api/58731/

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-api/58774/

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
https://jenkins.ceph.com/job/ceph-api/58801/

**Comment by vshankar** [comment]
@rishabh-d-dave 

```
2023-07-25 17:41:25,714.714 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 712, in 
2023-07-25 17:41:25,714.714 INFO:__main__:    _validate_json(val[key], sub_schema, path + [key])
2023-07-25 17:41:25,714.714 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 710, in _validate_json
2023-07-25 17:41:25,714.714 INFO:__main__:    raise _ValError('unknown keys: {}'.format(unknown_keys), path)
2023-07-25 17:41:25,714.714 INFO:__main__:tasks.mgr.dashboard.helper._ValError: In `input['fs_map']['filesystems'][0]['mdsmap']['flags_state']`: unknown keys: {'refuse_standby_for_another_fs'}
2023-07-25 17:41:25,714.714 INFO:__main__:
2023-07-25 17:41:25,714.714 INFO:__main__:During handling of the above exception, another exception occurred:
```

The new option `refuse_standby_for_another_fs` needs to be updated in some dashboard test.

**Comment by trociny** [comment]
> The new option refuse_standby_for_another_fs needs to be updated in some dashboard test.

@vshankar thanks, I will update it and add the commit here.

**Comment by trociny** [comment]
@rishabh-d-dave @vshankar I believe the dashboard tests are fixed now (the ceph API tests passed).

**Comment by rishabh-d-dave** [comment]
> @rishabh-d-dave
> 
> ```
> 2023-07-25 17:41:25,714.714 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 712, in 
> 2023-07-25 17:41:25,714.714 INFO:__main__:    _validate_json(val[key], sub_schema, path + [key])
> 2023-07-25 17:41:25,714.714 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 710, in _validate_json
> 2023-07-25 17:41:25,714.714 INFO:__main__:    raise _ValError('unknown keys: {}'.format(unknown_keys), path)
> 2023-07-25 17:41:25,714.714 INFO:__main__:tasks.mgr.dashboard.helper._ValError: In `input['fs_map']['filesystems'][0]['mdsmap']['flags_state']`: unknown keys: {'refuse_standby_for_another_fs'}
> 2023-07-25 17:41:25,714.714 INFO:__main__:
> 2023-07-25 17:41:25,714.714 INFO:__main__:During handling of the above exception, another exception occurred:
> ```
> 
> The new option `refuse_standby_for_another_fs` needs to be updated in some dashboard test.

@trociny The CephFS integration tests ran fine but the CI job "Ceph API tests" fails consistently. Please fix this issue so that this PR can a move step towards getting merged.

**Comment by rishabh-d-dave** [comment]
> @rishabh-d-dave @vshankar I believe the dashboard tests are fixed now (the ceph API tests passed).

Yes, dashboard tests are fixed. We can merge this PR but I guess its better to get an approval from Dashboard team as well. I've messaged Ernesto on Slack requesting review for this PR.

**Review by epuertat — APPROVED** [review]
LGT @ceph/dashboard!

**Comment by rishabh-d-dave** [comment]
@vshankar 
> @rishabh-d-dave good to merge?

I am waiting to discuss some failures with you before merging this PR.

**Comment by vshankar** [comment]
> @vshankar 
> > @rishabh-d-dave good to merge?
> 
> I am waiting to discuss some failures with you before merging this PR.

According to https://github.com/ceph/ceph/pull/51942#issuecomment-1653481476 this is ready to merge, isn't it?

**Comment by rishabh-d-dave** [comment]
> > @vshankar
> > > @rishabh-d-dave good to merge?
> > 
> > 
> > I am waiting to discuss some failures with you before merging this PR.
> 
> According to [#51942 (comment)](https://github.com/ceph/ceph/pull/51942#issuecomment-1653481476) this is ready to merge, isn't it?

Right but since there were too many re-runs, I included it in one of batches for testing again.

**Comment by vshankar** [comment]
> Right but since there were too many re-runs, I included it in one of batches for testing again.

As discussed - pending RCA for two test failures before merging.

**Review by rishabh-d-dave — APPROVED** [review]
CephFS integration tests ran fine - https://tracker.ceph.com/projects/cephfs/wiki/Main#6-Sep-2023.

**Comment by ljflores** [comment]
Hey @rishabh-d-dave, I think this caused a compilation failure. See https://tracker.ceph.com/issues/62729 for details

**Comment by ljflores** [comment]
@trociny FYI

**Pull request merged**
